images/tensorflow-notebook/cuda12/Dockerfile

```diff
@@ -0,0 +1,34 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+ARG REGISTRY=quay.io
+ARG OWNER=jupyter
+ARG BASE_CONTAINER=$REGISTRY/$OWNER/scipy-notebook
+FROM $BASE_CONTAINER
+
+LABEL maintainer="Jupyter Project <jupyter@googlegroups.com>"
+
+# Fix: https://github.com/hadolint/hadolint/wiki/DL4006
+# Fix: https://github.com/koalaman/shellcheck/wiki/SC3014
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+
+# Install CUDA libs and cuDNN with mamba
+RUN mamba install --yes -c nvidia/label/cuda-12.3.2 \
+    'cuda-libraries-dev' \
+    'cuda-nvcc' \
+    'cuda-nvtx' \
+    'cuda-cupti' \
+    'cudnn' && \
+    mamba clean --all -f -y && \
+    fix-permissions "${CONDA_DIR}" && \
+    fix-permissions "/home/${NB_USER}"
+
+ENV XLA_FLAGS="--xla_gpu_cuda_data_dir=${CONDA_DIR}" \
+    LD_LIBRARY_PATH="${CONDA_DIR}/lib/${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}" \
+    # https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/docker-specialized.html#dockerfiles
+    NVIDIA_VISIBLE_DEVICES="all" \
+    NVIDIA_DRIVER_CAPABILITIES="compute,utility"
+
+# Install Tensorflow with pip
+RUN pip install --no-cache-dir tensorflow && \
```

Comment:
Maybe we can use tensorflow from mamba?
I think in such a case we won't even need to list the dependencies.

Reply:
> * The current version of `tensorflow-gpu` on [conda-forge](https://anaconda.org/conda-forge/tensorflow-gpu/files) is 2.15.0 and has a CUDA 11.8 build. So if you want a cuda11 variant, I can try to use the conda-forge version for that, but the TensorFlow version is not up-to-date.

This is the kind of complexity I recall when getting a tensorflow gpu image working, where the conda-forge version is often outdated, and trying to install with `pip` for gpu support was complicated.

I think if we choose between install complexity or relying on something regularly outdated, the install complexity may be prefered - otherwise we introduce things we can't control. At the same time, the fact that its outdated etc relates to how complicated it may be to keep installing something that works over time, which we then may be taking on.

This PR will probably demonstrate the current maturity of tensorflow gpu stuff upstream, if its as bad as I experienced it was a while back, then I think its better to not try to support this.

Reply:
Yes, it was very complicated and the conda-forge `tensorflow-gpu` package helped by providing `cudatoolkit` and `cudnn` within the same toolchain. But I just tried to install `tensorflow-gpu` from conda-forge into scipy-notebook and it fails due to conflicts. So maybe it changed and nowadays the installation of `tensorflow` from PyPI and `cuda` and `cudnn` from nvidia's conda-channel is the easiest way. For maintenance I imagine using the cuda version from the [tested builds](https://www.tensorflow.org/install/source#gpu) for a new TensorFlow release should work.

Reply:
In my opinion, we can't tell if the current maturity of tensorflow packages is better until we merge this PR, have weekly builds, and have several releases of cuda/cudnn and the tensorflow package itself.

So, I am ok with how Dockerfile currently looks like, but we need to see if it's gonna be ok after a few releases.

Reply:
I am ready to give it a try as a maintainer (and I can always disable the build just by changing a few lines in the `docker.yml` config-like file).